com.unity.render-pipelines.universal/Runtime/Decal/ShaderPassDecal.hlsl

```diff
@@ -0,0 +1,291 @@
+#endif
+#ifdef DECAL_SCREEN_SPACE
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl"
+#endif
+#ifdef DECAL_GBUFFER
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/UnityGBuffer.hlsl"
+#endif
+#ifdef DECAL_RECONSTRUCT_NORMAL
+#include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/NormalReconstruction.hlsl"
+#endif
+
+#include "Packages/com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/Varyings.hlsl" // VertMesh
+
+void MeshDecalsPositionZBias(inout Varyings input)
+{
+#if UNITY_REVERSED_Z
+    input.positionCS.z -= _DecalMeshDepthBias;
+#else
+    input.positionCS.z += _DecalMeshDepthBias;
+#endif
+}
+
+PackedVaryings Vert(Attributes inputMesh)
+{
+    Varyings varyingsType = (Varyings)0;
+#ifdef DECAL_MESH
+    if (_DecalMeshBiasType == DECALMESHDEPTHBIASTYPE_VIEW_BIAS) // TODO: Check performance of branch
+    {
+        float3 viewDirectionOS = GetObjectSpaceNormalizeViewDir(inputMesh.positionOS);
+        inputMesh.positionOS += viewDirectionOS * (_DecalMeshViewBias);
+    }
+    varyingsType = BuildVaryings(inputMesh);
+#if (SHADERPASS == SHADERPASS_DECAL_SCREEN_SPACE_MESH) || (SHADERPASS == SHADERPASS_DECAL_GBUFFER_MESH)
+    OUTPUT_LIGHTMAP_UV(inputMesh.uv1, unity_LightmapST, varyingsType.lightmapUV);
+    OUTPUT_SH(varyingsType.normalWS, varyingsType.sh);
+#endif
+    if (_DecalMeshBiasType == DECALMESHDEPTHBIASTYPE_DEPTH_BIAS) // TODO: Check performance of branch
+    {
+        MeshDecalsPositionZBias(varyingsType);
+    }
+#else
+    varyingsType = BuildVaryings(inputMesh);
+#endif
+    return PackVaryings(varyingsType);
+}
+
+void Frag(PackedVaryings packedInput,
+#if defined(DECAL_DBUFFER)
+    OUTPUT_DBUFFER(outDBuffer)
+#elif defined(DECAL_SCREEN_SPACE)
+    out half4 outColor : SV_Target0
+#elif defined(DECAL_GBUFFER)
+    out FragmentOutput fragmentOutput
+#elif defined(DECAL_FORWARD_EMISSIVE)
+    out half4 outEmissive : SV_Target0
+#elif defined(SCENEPICKINGPASS)
+    out float4 outColor : SV_Target0
+#else
+#error SHADERPASS_is_not_correctly_define
+#endif
+)
+{
+#ifdef SCENEPICKINGPASS
+    outColor = _SelectionID;
+#else
+    UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(packedInput);
+    UNITY_SETUP_INSTANCE_ID(packedInput);
+    Varyings input = UnpackVaryings(packedInput);
+
+    DecalSurfaceData surfaceData;
+    float angleFadeFactor = 1.0;
+
+#ifdef DECAL_PROJECTOR
+    float depth = LoadSceneDepth(input.positionCS.xy);
+    PositionInputs posInput = GetPositionInput(input.positionCS.xy, _ScreenSize.zw, depth, UNITY_MATRIX_I_VP, UNITY_MATRIX_V);
+
+    // Transform from relative world space to decal space (DS) to clip the decal
+    float3 positionDS = TransformWorldToObject(posInput.positionWS);
+    positionDS = positionDS * float3(1.0, -1.0, 1.0);
+
+    // call clip as early as possible
+    float clipValue = 0.5 - Max3(abs(positionDS).x, abs(positionDS).y, abs(positionDS).z);
+    clip(clipValue);
```

Comment:
Caution: we have see that for HDRP on Metal early clip wasn't working, it was causing issue on texture due to derivative badly calculated (see the code in HDRP)

Reply:
Yea I saw the comment, if I understood correctly it was messing up ddx and ddy intstruction results? If yes, maybe I should move normal calculation above it.